packages/loopring_v3/contracts/amm/libamm/AmmWithdrawal.sol

```diff
@@ -0,0 +1,169 @@
+            AmmUtil.transferOut(token, amounts[i], msg.sender);
+        }
+    }
+
+    function withdrawInShutdown(
+        AmmData.State storage S
+        )
+        public
+    {
+        uint burnAmount = S.balanceOf[msg.sender];
+        require(burnAmount > 0, "INVALID_POOL_AMOUNT");
+
+        IExchangeV3 exchange = S.exchange;
+        bool withdrawalMode = exchange.isInWithdrawalMode();
+
+        uint size = S.tokens.length;
+        for (uint i = 0; i < size; i++) {
+            address token = S.tokens[i].addr;
+
+            require(
+                withdrawalMode && exchange.isWithdrawnInWithdrawalMode(S.accountID, token) ||
+                !withdrawalMode && !exchange.isForcedWithdrawalPending(S.accountID, token),
+                "PENDING_WITHDRAWAL"
+            );
+
+            // Check that nothing is withdrawable anymore.
+            require(
+                exchange.getAmountWithdrawable(address(this), token) == 0,
+                "MORE_TO_WITHDRAWAL"
+            );
+        }
+
+        // Withdraw proportionally to the liquidity owned, including the pool token itself
+        for (uint i = 0; i < size; i++) {
+            address token = S.tokens[i].addr;
+
+            uint balance = token == address(0) ?
+                address(this).balance :
+                ERC20(token).balanceOf(address(this));
+
+            // Withdraw the part owned by the pool
+            uint amount = balance
+                .sub(S.withdrawableBeforeShutdown[token])
+                .mul(burnAmount) / S.totalSupply;
+
+            AmmUtil.transferOut(token, amount, msg.sender);
+        }
+
+        S.burn(msg.sender, burnAmount);
+    }
+
+    function getWithdrawableAmounts(
+        AmmData.State storage S,
+        address               owner
+        )
+        internal
+        view
+        returns (uint, uint96[] memory)
+    {
+        uint size = S.tokens.length;
+
+        uint newJoinLocksStartIdx = S.joinLocksStartIdx[owner];
+        uint96[] memory amounts = new uint96[](size);
+
+        for (uint i = 0; i < size; i++) {
+            amounts[i] = S.withdrawable[S.tokens[i].addr][owner];
+        }
+
+        AmmData.TokenLock[] storage joinLocks = S.joinLocks[owner];
+        for (uint i = newJoinLocksStartIdx; i < joinLocks.length; i++) {
+            AmmData.TokenLock storage lock = joinLocks[i];
+            uint validUntil = S.approvedTx[lock.txHash];
+
+            if (validUntil <= block.timestamp) {
+                newJoinLocksStartIdx++;
+                if (lock.amounts.length == 1) {
+                    amounts[size - 1] = amounts[size - 1].add(lock.amounts[0]);
+                } else if (lock.amounts.length == size - 1) {
+                    for (uint j = 0; j < size - 1; j++) {
+                        amounts[j] = amounts[j].add(lock.amounts[j]);
+                    }
+                }
+            } else {
+                return (newJoinLocksStartIdx, amounts);
+            }
+        }
```

Comment:
I assume this needs to also check the exitLocks, because the case `lock.amounts.length == 1` doesn't happen for joins.